### ISSUE TITLE
Shared module use can prevent exceptions being caught

### ISSUE THREAD

**Issue by djbarker** [ninia/jep#111]
The `pandas` module seems to load the `pytz` module as a dependency, and some operations can raise a `pytz.NonExistentTimeError`. 

It appears that loading `pandas` as  shared module means that Python is unable to find that the `NonExistentTimeError` coming from `pandas` is the same as the one I've imported in my script to put in the `except` clause. If I do not use shared modules the code runs as expected.

A concrete example:

```java
import jep.Jep;
import jep.JepConfig;

public class ExceptionTypeTest {

    public static void main(String[] args) throws Exception {

        JepConfig jepConfig = new JepConfig();
        jepConfig.setInteractive(false);

        jepConfig.addSharedModules("pandas");  // comment out this line to see the expected print statements

        String script1 =
                        "import pandas as pd\n" +
                        "import pytz\n" +
                        "import sys\n" +
                        "d = pd.to_datetime('2017-10-15 00:00')\n";

        String script2 =
                        "try:\n" +
                        "\td.tz_localize('America/Sao_Paulo');\n" +  // this line is only caught properly without pandas as a shared module
                        "\traise pytz.NonExistentTimeError();\n" +   // this line is caught properly in both situations
                        "except pytz.NonExistentTimeError as e:\n" +
                        "\tprint('Saw exception');\n" +
                        "\tprint(type(e), e)\n" +
                        "\tsys.stdout.flush()\n";

        Jep jep = new Jep(jepConfig);
        for(String line : script1.split("\n")) {
            jep.eval(line);
        }
        jep.eval(script2);
        jep.close();
    }

}
```

I suppose this is a symptom of a more fundamental underlying issue with shared modules. 

- OS Platform, Distribution, and Version: Linux, Fedora 26
- Python Distribution and Version: Python 3.6.4, Anaconda
- Java Distribution and Version: jdk 1.8

**Comment by bsteffensmeier** [comment]
There is more than one pytz module, one in your jep interpreter and one coming over with pandas from the shared main interpreter. If you add pytz to the list of shared modules then it should work. I don't see anything else we can do. 

You mention there may be fundamental underlying issue with shared modules and I agree. Shared modules was implemented as a last resort work-around for extension modules that won't support sub-interpreters. Shared modules will never be an elegant solution, and it is troublesome that once you start using them it seems to suck in more and more modules but I can't think of any way to stop it.

Sub-interpreters are a documented part of the C-extension API for python and other projects such as mod_wsgi try to use sub-interpreters but they run into problems with native extensions, similar to what we see on jep without shared module. In my opinion, the real underlying issue is with the python ecosystem, module writers will not write their extensions to support sub-interpreters, this makes it very difficult to actually use sub-interpreters, which means they aren't used much, and since sub-interpreters aren't used much module writers have no reason to support them.

The only elegant solution I can come up with would be for all python extension writers to support using their modules in sub-interpreters. I don't know what issues caused pandas to be shared, it may be a bug in pandas or it may have been one of their dependencies. I would recommend you try to figure out why pandas cannot be used safely in sub-interpreters and write issues with the projects that are causing problems asking them to support sub-interpreters.

**Comment by djbarker** [comment]
Hi Ben.

Yes, this is what I thought was was happening. I feared this might be your answer. In my specific case adding `pytz` as shared does make my code run.

In general this is not very extensible, however, since it requires one to know which modules will be imported by each shared module ahead of time (before even starting the interpreter). But in principle these shared modules could load any different module, even dependent on what happens at run-time.

This means that as soon as we have shared modules (an absolute must to get `numpy` working with multiple `Jep` instances) we can no longer be truly sure our `except Foo:` statements will actually work.

- Dan


_I agree, it's very frustrating that sub-interpreters are a documented feature but seem to be so universally ignored (or just forgotten) by extension developers._
